src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -1577,25 +1577,288 @@ class QuantizedOr(QuantizedOpUnivariateOfEncrypted, QuantizedOp):
+            stats=self.output_quant_stats,
+            params=self.output_quant_params,
+        )
+
+    def dump_dict(self) -> Dict:
+        metadata = super().dump_dict()
+        if hasattr(self, "min_non_zero_value"):
+            metadata["min_non_zero_value"] = self.min_non_zero_value
+        if hasattr(self, "divider_quantizer"):
+            metadata["divider_quantizer"] = self.divider_quantizer
+        return metadata
+
+    @staticmethod
+    def load_dict(metadata: Dict):
+        obj = super(QuantizedDiv, QuantizedDiv).load_dict(metadata)
+        if "min_non_zero_value" in metadata:
+            obj.min_non_zero_value = metadata.pop("min_non_zero_value")
+        if "divider_quantizer" in metadata:
+            obj.divider_quantizer = metadata.pop("divider_quantizer")
+        return obj
+
+    def can_fuse(self) -> bool:
+        """Determine if this op can be fused.
+
+        Add operation can be computed in float and fused if it operates over inputs produced
+        by a single integer tensor.
+
+        Returns:
+            bool: Whether the number of integer input tensors allows computing this op as a TLU
+        """
+
+        return len(self._int_input_names) == 1
+
 
-class QuantizedMul(QuantizedOpUnivariateOfEncrypted, QuantizedOp):
-    """Multiplication operator.
+class QuantizedMul(QuantizedMixingOp):
+    """Quantized Multiplication operator.
 
-    Only multiplies an encrypted tensor with a float constant for now. This operation will
-    be fused to a (potentially larger) TLU.
+    Can multiply either two variables (both encrypted) or a variable and a constant
     """
 
     _impl_for_op_named: str = "Mul"
 
+    def q_impl(
+        self,
+        *q_inputs: ONNXOpInputOutputType,
+        calibrate_rounding: bool = False,
+        **attrs,
+    ) -> ONNXOpInputOutputType:
+
+        # If either input is a RawOpOutput or if the op can be fused,
+        # perform the op in the TLU using FP32
+        if (
+            len(q_inputs) == 1
+            or isinstance(q_inputs[0], RawOpOutput)
+            or isinstance(q_inputs[1], RawOpOutput)
+            or self.can_fuse()
+        ):
+            return super().q_impl(*q_inputs, **attrs)
+
+        # For mypy
+        assert self.output_quant_params is not None
+        assert self.output_quant_params.scale is not None
+        assert self.output_quant_params.zero_point is not None
+
+        prepared_inputs = self._prepare_inputs_with_constants(
+            *q_inputs, calibrate=False, quantize_actual_values=True
+        )
+
+        q_input_0: QuantizedArray = prepared_inputs[0]
+        q_input_1: QuantizedArray = prepared_inputs[1]
+
+        assert q_input_0.quantizer.scale is not None
+        assert q_input_0.quantizer.zero_point is not None
+
+        assert q_input_1.quantizer.scale is not None
+        assert q_input_1.quantizer.zero_point is not None
+
+        # Remove the manual encrypted multiplication when we
```

Comment:
so how's the following different than the `enc_mul` we already have ? same as above, couldn't the following be re-used from somewhere else (like `QuantizedGemm` or `enc_mul`) ?

Reply:
We could.

Reply:
same as above, worth an issue + fixme then